### PR TITLE
Fix comparison of payload addresses, handle empty address

### PR DESCRIPTION
**Pull request by turbolent**
Payloads may have an owner which is empty (global registers). Payload grouping incorrectly returned multiple groups for empty owners. This was caused by the assumption that addresses in payloads' encoded keys all have the same length.

Properly decode addresses from payloads using `PayloadToAddress`. This is inefficient, but fixes the incorrect grouping. We might want to only decode the owner key part (0) and avoid decoding the key key part (1).

### CODE REVIEW

**Comment by janezpodhostnik** [comment]
Is the fact that the grouping returns multiple groups for empty owners a problem anywhere? The empty address group(s) don't get fed into the account migrators anyway if I'm not mistaken.

Can we accept that there will be multiple groups for the empty address, and instead sort of join those together after the grouping?

`PayloadToAddress` is probably going to be quite a big hit here. If its needed, it might be better to to just compute the addresses beforehand.

**Comment by turbolent** [comment]
> Is the fact that the grouping returns multiple groups for empty owners a problem anywhere? The empty address group(s) don't get fed into the account migrators anyway if I'm not mistaken.

Currently / so far, all account groups generated by the grouping where processed, including the ones with a zero address / empty owner. Previously this did not cause problems, but in #5897 we started to index account groups by address, and the duplicates are a problem. 

However, as you mentioned, given that the zero address / empty owner registers are not actually for a real account, and no Cadence values are stored in them (right?), we can probably just skip the groups with a zero address? I'll try to just skip them and see.

> `PayloadToAddress` is probably going to be quite a big hit here. 

Yeah, it definitely does more than we need. I'd still prefer using a correct address extraction function, the `encodedKeyAddressPrefixLength` approach seems brittle in general, given that encoding of payload keys is implemented elsewhere and may change at any time, silently breaking this assumption.

> If its needed, it might be better to to just compute the addresses beforehand.
Can you elaborate? Would be great to optimize this, just not quite sure what you mean.

**Comment by janezpodhostnik** [comment]
> and no Cadence values are stored in them (right?)

yes, they are all values handled by the FVM. And not cadence values

**Comment by turbolent** [comment]
I made handling the duplicate owner case more lenient in [94be933](https://github.com/onflow/flow-go/pull/5942/commits/94be93357d0f3e7990eb2dd1bb18bf02784cf03a): instead of failing, it now merges registers together, and logs a warning. This isn't strictly needed, but even works with the old sort function in the account grouping code that used `encodedKeyAddressPrefixLength`.

**Comment by fxamacker** [comment]
@turbolent this PR looks good to me.  Do you want to updated this PR to use `Payload.Address()` in PR #5948?

**Review by fxamacker — APPROVED** [review]

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 1]
```diff
@@ -488,10 +489,12 @@ func newByAccountRegistersFromPayloadAccountGrouping(
 		for accountRegisters := range results {
 			oldAccountRegisters := registersByAccount.SetAccountRegisters(accountRegisters)
 			if oldAccountRegisters != nil {
-				return fmt.Errorf(
-					"duplicate account registers for account %s",
-					accountRegisters.Owner(),
-				)
+				log.Warn().Msgf("account registers already exist for %x, merging", accountRegisters.Owner())
+
+				err := accountRegisters.Merge(oldAccountRegisters)
```
Comment (zhangchiqing) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:494]: Could we add some comments here to explain why we decide to merge in this case?

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 2]
```diff
@@ -488,10 +489,12 @@ func newByAccountRegistersFromPayloadAccountGrouping(
 		for accountRegisters := range results {
 			oldAccountRegisters := registersByAccount.SetAccountRegisters(accountRegisters)
 			if oldAccountRegisters != nil {
-				return fmt.Errorf(
-					"duplicate account registers for account %s",
-					accountRegisters.Owner(),
-				)
+				log.Warn().Msgf("account registers already exist for %x, merging", accountRegisters.Owner())
```
Comment (zhangchiqing) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:492]: maybe log the register count as well if useful.

**Review by zhangchiqing — APPROVED** [review]

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go (hunk 2 above)
Comment (turbolent) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go]: Done in [9adca6c](https://github.com/onflow/flow-go/pull/5942/commits/9adca6cf6eefe44e6918cbcc17164fe324e2d04a)

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go (hunk 1 above)
Comment (turbolent) [reply cmd/util/cmd/execution-state-extract/execution_state_extract.go:503]: Done in [9adca6c](https://github.com/onflow/flow-go/pull/5942/commits/9adca6cf6eefe44e6918cbcc17164fe324e2d04a)

**Pull request merged**